diff -ur a/configure.ac b/configure.ac
--- a/configure.ac	2010-04-02 15:28:11.000000000 +0100
+++ b/configure.ac	2015-01-11 11:23:28.883811698 +0000
@@ -23,6 +23,7 @@
 . $srcdir/version.sh
 gl_INIT_PACKAGE([gettext], [$VERSION_NUMBER])
 AM_INIT_AUTOMAKE([silent-rules])
+AC_PROG_CXX
 
 dnl Override automake's tar command used for creating distributions.
 am__tar='${AMTAR} chof - --owner=root --group=root "$$tardir"'
