diff -ur a/gettext-runtime/m4/ansi-c++.m4 b/gettext-runtime/m4/ansi-c++.m4
--- a/gettext-runtime/m4/ansi-c++.m4	2009-06-28 20:44:04.000000000 +0100
+++ b/gettext-runtime/m4/ansi-c++.m4	2015-01-11 12:09:30.633336749 +0000
@@ -11,7 +11,7 @@
 
 AC_DEFUN([gt_PROG_ANSI_CXX],
 [
-AC_CHECK_PROGS(CXX, $CCC g++ c++ gpp aCC CC cxx cc++ cl FCC KCC RCC xlC_r xlC, :)
+AC_CHECK_TOOLS(CXX, $CCC g++ c++ gpp aCC CC cxx cc++ cl FCC KCC RCC xlC_r xlC, :)
 if test "$CXX" != ":"; then
   dnl Use a modified version of AC_PROG_CXX_WORKS that does not exit
   dnl upon failure.
