diff -ur a/gettext-tools/libgettextpo/Makefile.gnulib b/gettext-tools/libgettextpo/Makefile.gnulib
--- a/gettext-tools/libgettextpo/Makefile.gnulib	2010-05-09 02:38:11.000000000 +0100
+++ b/gettext-tools/libgettextpo/Makefile.gnulib	2015-01-11 13:11:25.943063038 +0000
@@ -590,9 +590,9 @@
 ## begin gnulib module memchr
 
 
-EXTRA_DIST += memchr.c memchr.valgrind
+EXTRA_DIST += memchr.valgrind
 
-EXTRA_libgnu_la_SOURCES += memchr.c
+libgnu_la_SOURCES += memchr.c
 
 ## end   gnulib module memchr
 
